File: cpp/Mod_parse.cpp
#pragma once

Tuple2<Int, Int> findBoundary(List<Tuple2<String, String>> t, int start){
	Tuple2<String, String> first = get(t, start);
	String token = get21(first);
	String inhalt = get22(first);
	
	int po = 0;
	
	if(token == "VAR" || token == "TRUE" || token == "FALSE"){
		return tuple2(start, start);
	}else{ // = PO
		for(int i = start;;i++){
			token = get21(get(t, i));
			if(token == "PO")
				po++;
			else if(token == "PC")
				po--;

			if(po == 0)
				return tuple2(start, i);
		}	
	}
}

Tuple3<Tree, Map<Node, String>, Int> parseRec(List<Tuple2<String, String>> t, Tree tree, Node n, Map<Node, String> map, int id){
	Tuple2<String, String> first = get(t, 0);
	String token = get21(first);
	String inhalt = get22(first);
	
	
	if(token == "VAR"){
		map = setImage(map, n, inhalt);
		return tuple3(tree, map, id);
	}else if(token == "TRUE" || token == "FALSE"){
		map = setImage(map, n, token);
		return tuple3(tree, map, id);
	}else{ // = PO

		Tuple2<String, String> second = get(t, 1);
		token = get21(second);
		
		int endeErste;
		
		if(token == "NOT"){
			String rootToken = "NOT";
			map = setImage(map, n, rootToken);
			
			endeErste = 0;
			
			
		}else{//binär
			Tuple2<Int, Int> ersteFormel = findBoundary(t, 1);
			int startErste = get21(ersteFormel);
			endeErste = get22(ersteFormel);

			String rootToken = get21(get(t, endeErste+1));
			map = setImage(map, n, rootToken);

			Node newRoot1 = createNode(id);
			tree = addOwnLeaf(tree, n, newRoot1);
			Tuple3<Tree, Map<Node, String>, Int> teil1 = parseRec(sublist(t, startErste, (endeErste-startErste+1)), tree, newRoot1, map, id+1);

			tree = get31(teil1);
			map = get32(teil1);
			id = get33(teil1);
		}

		Tuple2<Int, Int> zweiteFormel = findBoundary(t, endeErste + 2);
		int startZweite = get21(zweiteFormel);
		int endeZweite = get22(zweiteFormel);
		
		Node newRoot2 = createNode(id);
		tree = addOwnLeaf(tree, n, newRoot2);
		Tuple3<Tree, Map<Node, String>, Int> teil2 = parseRec(sublist(t, startZweite, (endeZweite-startZweite+1)), tree, newRoot2, map, id+1);

		tree = get31(teil2);
		map = get32(teil2);
		id = get33(teil2);
	}
	return tuple3(tree, map, id);;
}

Tuple2<Tree, Map<Node, String>> parse(List<Tuple2<String, String>> t){
	Map<Node, String> map;
	int id = 1;
	Node root = createNode(id);
	Tree tree = newTreeOwn(root);
	id++;
	
	Tuple3<Tree, Map<Node, String>, Int> rec = parseRec(t, tree, root, map, id);
	return tuple2(get31(rec), get32(rec));
	
	
}

File: hs/Mod_parse.hs
module Mod_parse ( parse ) where
import Prelude hiding (and, or)
import Debug.Trace
import Splib
import Types

import Mod_treeToString


findBoundary :: [(String, String)] -> Int -> (Int, Int)
findBoundary t start
 | token == "VAR" || token == "TRUE" || token == "FALSE" = (start, start)
 | otherwise = findBoundaryHelper t start (start+1) 1
  where
   first = get t start
   token = get21 first

findBoundaryHelper :: [(String, String)] -> Int -> Int -> Int -> (Int, Int)
findBoundaryHelper t start i po
 | po == 0 = (start, (i-1))
 | token == "PO" = (findBoundaryHelper t start (i+1) (po+1))
 | token == "PC" = (findBoundaryHelper t start (i+1) (po-1))
 | otherwise = (findBoundaryHelper t start (i+1) po)
  where
   token = get21 (get t i)

parseRec :: [(String, String)] -> Tree -> Node -> (Map Node String) -> Int -> (Tree, (Map Node String), Int)
parseRec t tree n mapParam idParam 
 | token == "VAR" = (tree, (setImage mapParam n inhalt), idParam)
 | token == "TRUE" || token == "FALSE" = (tree, (setImage mapParam n token), idParam)
 | otherwise = (parseRecElse t tree n mapParam idParam)
 where 
  first = get t 0
  token = get21 first
  inhalt = get22 first
  
parseRecElse t tree n mapParam idParam
 | token == "NOT" = (parseRecFinish t tree n (setImage mapParam n "NOT") idParam 0)
 | otherwise = (parseRecFirst t tree n mapParam idParam)
  where 
   second = get t 1
   token = get21 second
   
parseRecFirst t tree n mapParam idParam = (parseRecFinish t newnewTree n newnewMap newnewID endeErste)
 where
  ersteFormel = findBoundary t 1
  startErste = get21 ersteFormel
  endeErste = get22 ersteFormel
  rootToken = get21 (get t (endeErste+1))
  newMap = setImage mapParam n rootToken
  newRoot1 = createNode idParam
  newTree = addOwnLeaf tree n newRoot1
  teil1 = parseRec (sublist t startErste (endeErste-startErste+1)) newTree newRoot1 newMap (idParam+1)
  newnewTree = get31 teil1
  newnewMap = get32 teil1
  newnewID = get33 teil1
  
parseRecFinish t tree n mapParam idParam  endeErste
 = (parseRec (sublist t startZweite (endeZweite-startZweite+1)) newTree newRoot2 mapParam (idParam+1))
 where
  zweiteFormel = findBoundary t (endeErste+2)
  startZweite = get21 zweiteFormel
  endeZweite = get22 zweiteFormel
  newRoot2 = createNode idParam
  newTree = addOwnLeaf tree n newRoot2

parse :: [(String, String)] -> (Tree, (Map Node String))
parse t = (get31 rec, get32 rec)
 where 
  rec = parseRec t (newTreeOwn root) root emptyMap 2
  root = createNode 1
  